File: 프로그래머스/3/43163. 단어 변환/단어 변환.kt
class Solution {
    var ans = 51
    lateinit var visit: BooleanArray
    
    fun solution(begin: String, target: String, words: Array<String>): Int {
        visit = BooleanArray(words.size) { false }
        dfs(begin, 0, target, words)
        return if (ans == 51) 0 else ans
    }
    
    fun dfs(curWord: String, depth: Int, target: String,  words: Array<String>) {
        if (curWord == target) {
            ans = minOf(ans, depth)
        }
        
        words.forEachIndexed { idx, word ->
            if (!visit[idx] && curWord.isOneDiff(word)) {
                visit[idx] = true
                dfs(word, depth + 1, target, words)
                visit[idx] = false
            }
        }
    }
    
    // 하나의 알파벳만 다른지 판별하는 함수
    fun String.isOneDiff(other: String): Boolean {
        var cnt = 0
        this.forEachIndexed { idx, _ ->
            if (this[idx] != other[idx]) cnt++
            if (cnt > 1) return false
        }
        return cnt == 1
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 1e9;
bool visited[51];
vector<string> word_list;

int diff(string str1, string str2) {
    int diff = 0;
    for (int i = 0; i < str1.length(); i++) {
        if (str1[i] != str2[i]) ++diff;
    }
    return diff;
}

void dfs(string cur, string target, int depth) {
    if (diff(cur, target) == 0) {
        if (answer > depth - 1) answer = depth - 1;
    }
    
    for (int i = 0; i < word_list.size(); i++) {
        if (visited[i]) continue;
        
        if (diff(cur, word_list[i]) == 1) {            
            visited[i] = true;
            dfs(word_list[i], target, depth + 1);
            visited[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    word_list = words;
    
    for (int i = 0; i < words.size(); i++) {
        visited[i] = true;
        dfs(begin, target, 1);
        visited[i] = false;
    }
    
    return answer == 1e9 ? 0 : answer;
}
